File: aspnet-in-practice/15/CS/VirtualPathProvider/App_Code/DatabaseFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Hosting;
using System.IO;

namespace ASPNET4InPractice
{
	public class DatabaseFile : VirtualFile
	{
		public DatabaseFile(string virtualPath) : base(virtualPath) { }

		public byte[] LastModifiedTimeStamp
		{
			get
			{
				return Utility.GetLastModifiedTimeStamp(VirtualPathUtility.ToAppRelative(VirtualPath));
			}
		}

		public override Stream Open()
		{
			// get file contents and write to the stream
			string fileContents = Utility.GetFileContents(VirtualPathUtility.ToAppRelative(VirtualPath));
			Stream stream = new MemoryStream();
			if (!string.IsNullOrEmpty(fileContents))
			{
				StreamWriter writer = new StreamWriter(stream);
				writer.Write(fileContents);
				writer.Flush();
				stream.Seek(0, SeekOrigin.Begin);
			}
			return stream;
		}
	}
}

File: aspnet-in-practice/15/VB/VirtualPathProvider/App_Code/DatabaseFile.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Hosting
Imports System.IO

Namespace ASPNET4InPractice
	Public Class DatabaseFile
		Inherits VirtualFile
		Public Sub New(ByVal virtualPath As String)
			MyBase.New(virtualPath)
		End Sub

		Public ReadOnly Property LastModifiedTimeStamp() As Byte()
			Get
				Return Utility.GetLastModifiedTimeStamp(VirtualPathUtility.ToAppRelative(VirtualPath))
			End Get
		End Property

		Public Overloads Overrides Function Open() As Stream
			' get file contents and write to the stream
			Dim fileContents As String = Utility.GetFileContents(VirtualPathUtility.ToAppRelative(VirtualPath))
			Dim stream As Stream = New MemoryStream()
			If Not String.IsNullOrEmpty(fileContents) Then
				Dim writer As New StreamWriter(stream)
				writer.Write(fileContents)
				writer.Flush()
				stream.Seek(0, SeekOrigin.Begin)
			End If
			Return stream
		End Function
	End Class
End Namespace